### PR TITLE
Active Learning: generic SubsetDatasetBuilder to support more datasets

### PR DESCRIPTION
**Pull request by BlackHC**
Rewrite the CIFAR10Subset into a SubsetDatasetBuilder that creates a unique id using "tfds_id", which works with any dataset and not only CIFAR-10 and CIFAR-100 ("id" is specific to these two datasets).

Tested with Places365 and CIFAR-100.

Updated tests to verify MNIST and CIFAR-10 both work. MNIST also lacks the "id" field.

### CODE REVIEW

File: baselines/jft/al_utils_test.py [hunk 1]
```diff
@@ -24,15 +22,38 @@
 
 class AlUtilsTest(tf.test.TestCase):
 
+  def test_mnist_subset_has_ids(self):
+    # NOTE: MNIST has not id field, so this tests the enumerate code path.
```
Comment (y0ast) [inline baselines/jft/al_utils_test.py:26]:
```suggestion
    # NOTE: MNIST has no id field, so this tests the enumerate code path.
```

File: baselines/jft/al_utils_test.py [hunk 2]
```diff
@@ -24,15 +22,38 @@
...
+    all_ids = []
+    n = 0
+    for example in dataset.as_numpy_iterator():
+      n += example['image'].shape[0]
+      all_ids.extend(example['id'])
+
+    all_ids.sort()
+
+    self.assertEqual(n, 60000)
+    self.assertEqual(all_ids, list(range(60000)))
```
Comment (y0ast) [inline baselines/jft/al_utils_test.py:46]: Would be nice to verify something about ids in test/train splits here too

File: baselines/jft/experiments/vit_l32_active_learning.py [hunk 3]
```diff
@@ -43,6 +43,7 @@ def get_config():
   config.test_split = ''  # set in sweep
   config.val_split = ''  # set in sweep
   config.train_split = ''  # set in sweep
+  config.test_split = 'test'
```
Comment (y0ast) [inline baselines/jft/experiments/vit_l32_active_learning.py:46]: @zi-w this line would be good to double check. I know there's been some discussion around the "right" test split.

File: baselines/jft/al_utils.py [hunk 4]
```diff
@@ -12,81 +12,134 @@
...
-        # jax prefers int32
-        # https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision
-        int_id = np.int32(int_id % np.iinfo(np.int32).max)
-
-      if subset_ids is not None:
-        # Hard fail on type errors
-        assert all(map(lambda id: isinstance(id, int), subset_ids))
+    if subset_ids is not None:
+      # Hard fail on type errors
+      assert all(map(lambda id: isinstance(id, int), subset_ids))
```
Comment (y0ast) [inline baselines/jft/al_utils.py:72]: This is probably a bit too defensive, we would notice it if it wasn't true.

File: baselines/jft/al_utils.py [hunk 5]
```diff
@@ -12,81 +12,134 @@
...
+        _subset_generator(
+            dataset=dataset,
+            dataset_info=self.base_dataset_builder.info,
+            subset_ids=self.subset_ids,
+            splitwise_id=True),
         output_signature=element_spec,
-    ).cache()
+    )
+
+    # This is a bit more complex: potentially cache before or after calling .shuffle.
```
Comment (y0ast) [inline baselines/jft/al_utils.py:126]: I don't follow the below, did you base it on some other code? Would be good to reference that here.

File: baselines/jft/al_utils.py [hunk 6]
```diff
@@ -12,81 +12,134 @@
...
+  def as_dataset(self,
+                 split: Union[str, tfds.core.ReadInstruction],
+                 *,
+                 shuffle_files: bool = False,
+                 read_config: Optional[tfds.ReadConfig] = None,
+                 **kwargs) -> tf.data.Dataset:
+    # We don't allow an empty split by virtue of the parameter declaration, so we always have a split.
+    read_config = dataclasses.replace(
+        kwargs.pop('read_config', tfds.ReadConfig()))
+    # Add the 'tfds_id' key to the samples which we can then parse.
```
Comment (y0ast) [inline baselines/jft/al_utils.py:106]:
```suggestion
    # Add the 'tfds_id' key to the samples which we can then parse.
    # From: https://www.tensorflow.org/datasets/api_docs/python/tfds/ReadConfig
```

**Review by y0ast — COMMENTED** [review]
Nice improvement to the dataset creation setup! Looks much less hacky. Is there any way we can upstream some of these changes? Would a PR to clu make our lives much easier?

Open question if we want to merge the cifar100 and places365 configs.

File: baselines/jft/al_utils.py (hunk 4 above)
Comment (BlackHC) [reply baselines/jft/al_utils.py:72]: I think it's good to fail really hard here because there are multiple code paths downstream and if a few non-ints snuck through it would be near impossible to find out otherwise.

Likewise, given that we don't expect to collect more than 200-300 samples, the runtime cost is manageable.

An alternative would be to add an 'acquire' method that encapsulates subset_ids away and have the check there only once.

File: baselines/jft/al_utils.py (hunk 5 above)
Comment (BlackHC) [reply baselines/jft/al_utils.py:126]: It's handwritten and follows the logic of reshuffling on every iteration or not. If we call cache before the shuffle, each epoch shuffle will be executed. Otherwise, if we call cache after shuffle, it will fix the first shuffle in place (in line with the behavior of that readconfig).

File: baselines/jft/experiments/vit_l32_active_learning.py (hunk 3 above)
Comment (BlackHC) [reply baselines/jft/experiments/vit_l32_active_learning.py:46]: Ah okay, then test_split can be removed here again

**Review by zi-w — APPROVED** [review]
